src/miniscript/iter.rs

```diff
@@ -0,0 +1,703 @@
+    /// if any. Otherwise returns an empty `Vec`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public keys within AST use [`iter_pubkeys()`] function, for example
+    /// `miniscript.iter_pubkeys().collect()`.
+    pub fn get_pubkeys(&self) -> Vec<PK> {
+        match self.node.clone() {
+            Terminal::PkK(key) => vec![key],
+            Terminal::Multi(_, keys) => keys,
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` with hashes of all public keys from the current miniscript item, if any.
+    /// Otherwise returns an empty `Vec`.
+    ///
+    /// For each public key the function computes hash; for each hash of the public key the function
+    /// returns it's cloned copy.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public key hashes within AST use [`iter_pubkey_hashes()`] function,
+    /// for example `miniscript.iter_pubkey_hashes().collect()`.
+    pub fn get_pubkey_hashes(&self) -> Vec<PK::Hash> {
+        match self.node.clone() {
+            Terminal::PkH(hash) => vec![hash],
+            Terminal::PkK(key) => vec![key.to_pubkeyhash()],
+            Terminal::Multi(_, keys) => keys.iter().map(PK::to_pubkeyhash).collect(),
+            _ => vec![],
+        }
+    }
+
+    /// Returns `Vec` of [PubkeyOrHash] entries, representing either public keys or public key
+    /// hashes, depending on the data from the current miniscript item. If there is no public
+    /// keys or hashes, the function returns an empty `Vec`.
+    ///
+    /// NB: The function analyzes only single miniscript item and not any of its descendants in AST.
+    /// To obtain a list of all public keys or hashes within AST use [`iter_pubkeys_and_hashes()`]
+    /// function, for example `miniscript.iter_pubkeys_and_hashes().collect()`.
+    pub fn get_pubkeys_and_hashes(&self) -> Vec<PubkeyOrHash<PK>> {
+        use self::PubkeyOrHash::*;
+        match self.node.clone() {
+            Terminal::PkH(hash) => vec![HashedPubkey(hash)],
+            Terminal::PkK(key) => vec![PlainPubkey(key)],
+            Terminal::Multi(_, keys) => keys.into_iter().map(PlainPubkey).collect(),
+            _ => vec![],
+        }
+    }
+
+    /// Iterates over all public keys and public key hashes by traversing a complete Miniscript
+    /// AST; calls `processor` callback providing each of the found public key or hash and
+    /// replaces miniscript item with the new one constructed out of results returned from the
+    /// callback. The function returns a new Miniscript AST.
+    ///
+    /// If you need to fail the process due to error, return `None` value from the processor
+    /// and replacement will fail with [Error::PubkeyProcessorFailure] error.
+    pub fn replace_pubkeys_and_hashes<T>(&self, processor: &T) -> Result<Self, Error>
```

Comment:
We already have a function called `translate_pk` which does something similar but at a more generic level. You can transform a Miniscript<String> to Miniscript<publickey>. 
https://github.com/apoelstra/rust-miniscript/blob/3610c1d508781f25752eb8336e9f2e159973f3af/src/descriptor/mod.rs#L73

I don't think this adds any new functionality to what already possible and can be removed from this PR.

Reply:
Well, actually I was creating this because I needed this sort of functionality and `translate_pk` was not there yet... Seems like while this PR was waiting for review, somebody have added this part

Reply:
That API is present from the start of the codebase, but maybe the name is not intuitive and difficult to look for the end-user. I think we should add an example of the `translate_pk` functionality.